src/lean_spec/subspecs/containers/state/state.py

```diff
@@ -311,70 +311,77 @@ def process_block_header(self, block: Block) -> "State":
         AssertionError
             If any header check fails.
         """
+        # Validation
+        parent_header = self.latest_block_header
+        parent_root = hash_tree_root(parent_header)
+
         # The block must be for the current slot.
         assert block.slot == self.slot, "Block slot mismatch"
 
         # The block must be newer than the current latest header.
-        assert block.slot > self.latest_block_header.slot, "Block is older than latest header"
+        assert block.slot > parent_header.slot, "Block is older than latest header"
 
         # The proposer must be the expected validator for this slot.
         assert self.is_proposer(block.proposer_index), "Incorrect block proposer"
 
         # The declared parent must match the hash of the latest block header.
-        assert block.parent_root == hash_tree_root(self.latest_block_header), (
-            "Block parent root mismatch"
-        )
+        assert block.parent_root == parent_root, "Block parent root mismatch"
 
-        # Build a dictionary of field updates to apply in one copy operation.
-        updates: Dict[str, Any] = {}
-
-        # Cache the parent root locally for repeated use.
-        parent_root = block.parent_root
+        # State Updates
 
         # Special case: first block after genesis.
-        #
-        # Mark genesis as both justified and finalized.
-        if self.latest_block_header.slot == Slot(0):
-            updates["latest_justified"] = self.latest_justified.model_copy(
-                update={"root": parent_root}
-            )
-            updates["latest_finalized"] = self.latest_finalized.model_copy(
-                update={"root": parent_root}
-            )
-
-        # Create mutable copies to work with to avoid modifying the original object.
-        new_historical_hashes = list(self.historical_block_hashes)
-        new_historical_hashes.append(parent_root)
-
-        new_justified_slots = list(self.justified_slots)
-        new_justified_slots.append(Boolean(self.latest_block_header.slot == Slot(0)))
+        is_genesis_parent = parent_header.slot == Slot(0)
+        new_justified = (
```

Comment:
Super nit: avoid ambiguity from the name diff

```suggestion
        new_latest_justified = (
```